task/mgmt-gateway/src/mgs_handler.rs

```diff
@@ -152,4 +221,104 @@ impl SpHandler for MgsHandler {
+        if !self.reset_requested {
+            return Err(ResponseError::SysResetTriggerWithoutPrepare);
+        }
+
+        let jefe = task_jefe_api::Jefe::from(crate::JEFE.get_task_id());
+        jefe.request_reset();
+
+        // If `request_reset()` returns, something has gone very wrong.
+        panic!()
+    }
+}
+
+#[derive(Default)]
+struct UpdateBuffer {
+    total_length: usize,
+    bytes_written: usize,
+    current_block: ArrayVec<[u8; BLOCK_SIZE_BYTES]>,
+}
+
+impl UpdateBuffer {
+    fn ingest_chunk(
+        &mut self,
+        update_task: &Update,
+        chunk: &UpdateChunk,
+    ) -> Result<(), ResponseError> {
+        // Reject chunks that don't match our current progress.
+        if chunk.offset as usize != self.bytes_written {
+            return Err(ResponseError::UpdateInProgress {
+                bytes_received: self.bytes_written as u32,
+            });
+        }
+
+        // Reject chunks that would go past the total size we're expecting.
+        if self.bytes_written + chunk.chunk_length as usize > self.total_length
+        {
+            return Err(ResponseError::InvalidUpdateChunk);
+        }
+
+        let mut data = &chunk.data[..chunk.chunk_length as usize];
+        while !data.is_empty() {
+            let cap = self.current_block.capacity() - self.current_block.len();
+            assert!(cap > 0);
+            let to_copy = usize::min(cap, data.len());
+
+            let current_block_index = self.bytes_written / BLOCK_SIZE_BYTES;
+            self.current_block.extend_from_slice(&data[..to_copy]);
+            data = &data[to_copy..];
+            self.bytes_written += to_copy;
+
+            // If the block is full or this is the final block, send it to the
+            // update task.
+            if self.current_block.len() == self.current_block.capacity()
+                || self.bytes_written == self.total_length
+            {
+                update_task
+                    .write_one_block(current_block_index, &self.current_block)
+                    .map_err(|err| ResponseError::UpdateFailed(err as u32))?;
+                self.current_block.clear();
+            }
+        }
+
+        // Finalizing the update is implicit (we finalize if we just wrote the
+        // last block). Should we make it explict somehow? Maybe that comes with
+        // adding auth / code signing?
+        if self.bytes_written == self.total_length {
+            update_task
+                .finish_image_update()
+                .map_err(|err| ResponseError::UpdateFailed(err as u32))?;
+            ringbuf_entry!(Log::UpdateComplete);
+        } else {
+            ringbuf_entry!(Log::UpdatePartial {
+                bytes_written: self.bytes_written
+            });
+        }
+
+        Ok(())
+    }
+}
+
+fn claim_update_buffer_static() -> &'static mut UpdateBuffer {
+    let update_buffer_array = mutable_statics! {
+        static mut BUF: [UpdateBuffer; 1] = [Default::default(); _];
```

Comment:
Add a TODO to explain why we're doing an array of 1 element here (since it's a `mutable_statics` limitation)